src/native/eventpipe/ep.c

```diff
@@ -1319,22 +1319,15 @@ ep_delete_provider (EventPipeProvider *provider)
 	// where we hold a provider after tracing has been disabled.
 	bool wait_for_provider_callbacks_completion = false;
 	EP_LOCK_ENTER (section1)
-		if (enabled ()) {
-			// Save the provider until the end of the tracing session.
-			ep_provider_set_delete_deferred (provider, true);
-
-			// The callback func must be previously set to null,
-			// otherwise callbacks might never stop coming.
-			EP_ASSERT (provider->callback_func == NULL);
-
-			// Calling ep_delete_provider within a Callback will result in a deadlock
-			// as deleting the provider with an active tracing session will block
-			// until all of the provider's callbacks are completed.
-			if (provider->callbacks_pending > 0)
-				wait_for_provider_callbacks_completion = true;
-		} else {
-			config_delete_provider (ep_config_get (), provider);
-		}
+		// The callback func must be set to null,
+		// otherwise callbacks might never stop coming.
+		provider->callback_func = NULL;
```

Comment:
Any reason not to call `ep_provider_set_delete_deferred(provider,true)` here? Down below would then be:
```
if(!enabled()) {
    config_delete_provider (ep_config_get (), provider);
}
```

Reply:
Ah good point since `config_delete_provider` will unregister + free the provider, then unconditionally calling `ep_provider_set_delete_deferred` works.